cpp/src/rolling/grouped_rolling.cu

```diff
@@ -94,6 +93,109 @@ std::unique_ptr<column> grouped_rolling_window(table_view const& group_keys,
 
 namespace detail {
 
+/// Preceding window calculation functor.
+template <bool preceding_less_than_1>
```

Comment:
What is the advantage and disadvantage of having this template?

Reply:
Ah, right. This is to enable the `if constexpr()` in `operator()`:
```c++
    if constexpr (preceding_less_than_1) {  // where 1 indicates only the current row.
      auto group_end = _group_offsets_begin[group_label + 1];
      return thrust::maximum{}(_preceding_window, -(group_end - 1 - idx));
    } else { ... }
```
When this used to be a lambda, this used to be:
```c++
    if (preceding < 1) { ... } else { ... }
```
That if check would have run once per row in the window. Or worst case N**2 times, over the column. The template was so that this is checked once.

Reply:
Has there been a tangible benefit to runtime perf with this change?

Reply:
Sorry if the refactor isn't immediately obvious. This commit shows the refactor: 589f2a57c4a979c832cf733d030b51cdc097ac54

Reply:
Ah, I don't have a benchmark in place for this. 

I did assume that lifting this if check out would be beneficial.

Reply:
Thanks for accommodating, @divyegala.

I've measured how much longer `grouped_rolling.cu` takes to compile, with all the changes in this PR thrown in.
Old: 98s
New: 93s

It's taking less time to compile now.  This might have to do with my materializing the preceding/following columns, to simplify the iterators passed to `cudf::detail::rolling_window`.

Reply:
As an aside, the compile time for `rolling_fixed_window.cu` has reduced from 80s to 65s.

So @divyegala's instinct was right, to keep an eye on the compile times.  I think I've mitigated it for now.